lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+--     c2 = colorful.color("#99ccff")
+--     c3 = c:mix(c2, 0.4) -- mix c with c2, giving c 40% and c2 60%
+--     c:darken(0.25) -- darken 25%
+--     c:lighten(0.25) -- lighten 25%
+--     c:saturate(0.1) -- saturate 10%
+--     c:desaturate(0.1) -- desaturate 10%
+--     c:grayscale() -- desaturate color completely to grayscale
+--     c:invert() -- inverts color
+--     c:complement() -- the compliment of the color (180deg hue shift)
+--
+-- If you want to create a color scheme you can use the color harmony functions:
+--     local left, right = c:analogous()
+--     local left, right = c:split_complementary()
+--     local left, right = c:triadic()
+--     local complement, analogous, analogous_complement = c:tetradic()
+--
+-- @author Kevin Zander
+-- @copyright 2019 Kevin Zander
+-- @module beautiful.colorful
+---------------------------------------------------------------------------
+
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+local colorful = {}
+
+local color = {
+    -- set the from and as tables
+    from = {},
+    as = {}
+}
+
+-- Metatable for color instances. This is set up near the end of the file.
+local color_mt = {}
+
+-- {{{ Helper functions
+
+-- Lua 5.1 does not support bit shifting operators
+local function lshift(x, disp)
+    return (x * 2^disp) % 2^32
+end
+
+local function rshift(x, disp)
+    return math.floor(x % 2^32 / 2^disp)
+end
+
+-- Do we want to pull in gears.math.round?
+local function round(x)
+    local d,_ = math.modf(x + (x >= 0 and 1 or -1) * 0.5)
+    return d
+end
+
+-- Clamp a number between a low and high
+local function clamp(v, l, h)
+    if v < l then return l end
+    if v > h then return h end
+    return v
+end
+
+-- Return true/false if a number is between a low and high
+local function between(v, l, h)
+    if l <= v and v <= h then return true end
+    return false
+end
+
+-- Expects Hue [0,360], Saturation [0,1], Lightness [0,1]
+-- Does NOT check bounds
+local function hsl_to_rgb(hue, saturation, lightness)
+    -- If s = 0 then it's a grey
+    if saturation == 0 then
+        return unpack({lightness, lightness, lightness})
+    end
+    local C = (1 - math.abs(2 * lightness - 1)) * saturation
+    local X = C * (1 - math.abs((hue / 60) % 2 - 1))
+    local m = lightness - (C / 2)
+    local r,g,b = 0,0,0
+    if 0 <= hue and hue <= 60 then
+        r,g,b = C,X,0
+    elseif 60 <= hue and hue <= 120 then
+        r,g,b = X,C,0
+    elseif 120 <= hue and hue <= 180 then
+        r,g,b = 0,C,X
+    elseif 180 <= hue and hue <= 240 then
+        r,g,b = 0,X,C
+    elseif 240 <= hue and hue <= 300 then
+        r,g,b = X,0,C
+    elseif 300 <= hue and hue <= 360 then
+        r,g,b = C,0,X
+    end
+    return unpack({r+m, g+m, b+m})
+end
+
+-- Expects Red, Blue, and Green [0,1]
+-- Does NOT check bounds
+local function rgb_to_hsl(red, green, blue)
+    local h,s,l = 0,0,0 -- luacheck: ignore
```

Comment:
Why the `luacheck: ignore`? I guess it is warning about unused values that are assigned here and in fact it is right. What's wrong with just writing `local h,s,l` without an assignment?

Reply:
Yeah luacheck was complaining about it for some reason, so I put it there. I think I tried removing the assignment and the unused error was still there. I'll try and see what it says (I just recently updated luacheck to the newest version from 0.19, this was done before that).